gridstatus/caiso.py

```diff
@@ -342,29 +969,73 @@ def _add_forecast_publish_time(
-            future_forecasts_publish_time = todays_publish_time - pd.Timedelta(days=1)
-
-        # Forecasts tomorrow and later get the future forecasts publish time
-        # Forecasts today and earlier get a publish time of the previous day at the
-        # publish time offset
-        data["Publish Time"] = np.where(
-            data["Interval Start"].dt.date > future_forecasts_publish_time.date(),
-            future_forecasts_publish_time,
-            data["Interval Start"].apply(
-                lambda x: x.floor("D").replace(hour=hour_offset, minute=minute_offset),
-            )
-            # DAM is for the next day
-            - pd.Timedelta(days=1),
-        )
+        match forecast_vintage:
+            case "DAM":
+                # If the current time is after the publish time, then future forecasts were
+                # published today. Otherwise, they were published in a previous day.
+                if current_time > todays_publish_time:
+                    future_forecasts_publish_time = todays_publish_time
+                else:
+                    future_forecasts_publish_time = todays_publish_time - pd.Timedelta(
+                        days=1,
+                    )
+
+                # Forecasts tomorrow and later get the future forecasts publish time
+                # Forecasts today and earlier get a publish time of the previous day at the
+                # publish time offset
+                data["Publish Time"] = np.where(
+                    data["Interval Start"].dt.date
+                    > future_forecasts_publish_time.date(),
+                    future_forecasts_publish_time,
+                    data["Interval Start"].apply(
+                        lambda x: x.floor("D").replace(
+                            hour=hour_offset,
+                            minute=minute_offset,
+                        ),
+                    )
+                    # DAM is for the next day
+                    - pd.Timedelta(days=1),
+                )
+            case "2DA":
+                data["Publish Time"] = data["Interval Start"].apply(
+                    lambda x: x.floor("D").replace(
+                        hour=hour_offset,
+                        minute=minute_offset,
+                    )
+                    - pd.Timedelta(days=2),
+                )
+            case "7DA":
+                data["Publish Time"] = data["Interval Start"].apply(
+                    lambda x: x.floor("D").replace(
+                        hour=hour_offset,
+                        minute=minute_offset,
+                    )
+                    - pd.Timedelta(days=7),
+                )
+            case "RTM5":
+                data["Publish Time"] = data["Interval Start"] - pd.Timedelta(minutes=5)
+            case "RTM15":
+                data["Publish Time"] = data["Interval Start"] - pd.Timedelta(minutes=15)
+            case "ACTUAL":
+                data["Publish Time"] = data["Interval Start"]
```

Comment:
Is "ACTUAL" the real recorded values? In that case, I don't think we need a publish time.

Reply:
If we are going to have a publish time, it should be +3 minutes after the interval.

<img width="1266" alt="Screenshot 2025-01-19 at 20 32 35" src="https://github.com/user-attachments/assets/4d89814a-549d-4395-8748-8eb799944762" />